src/backend/distributed/transaction/distributed_deadlock_detection.c

```diff
@@ -25,11 +27,355 @@
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * CheckDeadlockForDistributedTransaction does a DFS starting with the given
+ * transaction node and checks for a cycle (i.e., the node can be reached again
+ * while traversing the graph).
+ *
+ * Finding a cycle  indicates a distributed deadlock and the function returns
+ * true on that case. Also, the deadlockPath is filled with the transaction
+ * nodes that form the cycle.
+ */
+static bool
+CheckDeadlockForTransactionNode(TransactionNode *startingTransactionNode,
+								TransactionNode **transactionNodeStack,
+								List **deadlockPath)
+{
+	List *toBeVisitedNodes = NIL;
+	int currentStackDepth = 0;
+
+	/*
+	 * We keep transactionNodeStack to keep track of the deadlock paths. At this point,
+	 * adjust the depth of the starting node and set the stack's first element with
+	 * the starting node.
+	 */
+	transactionNodeStack[currentStackDepth] = startingTransactionNode;
+
+	PrependOutgoingNodesToQueue(startingTransactionNode, currentStackDepth,
+								&toBeVisitedNodes);
+
+	/* traverse the graph and search for the deadlocks */
+	while (toBeVisitedNodes != NIL)
+	{
+		QueuedTransactionNode *queuedTransactionNode =
+			(QueuedTransactionNode *) linitial(toBeVisitedNodes);
+		TransactionNode *currentTransactionNode = queuedTransactionNode->transactionNode;
+
+		toBeVisitedNodes = list_delete_first(toBeVisitedNodes);
+
+		/* cycle found, let the caller know about the cycle */
+		if (currentTransactionNode == startingTransactionNode)
+		{
+			BuildDeadlockPathList(queuedTransactionNode, transactionNodeStack,
+								  deadlockPath);
+
+			return true;
+		}
+
+		/* don't need to revisit the node again */
+		if (currentTransactionNode->transactionVisited)
+		{
+			continue;
+		}
+
+		currentTransactionNode->transactionVisited = true;
+
+		/* set the stack's corresponding element with the current node */
+		currentStackDepth = queuedTransactionNode->currentStackDepth;
+		transactionNodeStack[currentStackDepth] = currentTransactionNode;
+
+		PrependOutgoingNodesToQueue(currentTransactionNode, currentStackDepth,
+									&toBeVisitedNodes);
+	}
+
+	return false;
+}
+
+
+/*
+ * PrependOutgoingNodesToQueue prepends the waiters of the input transaction nodes to the
+ * toBeVisitedNodes.
+ */
+static void
+PrependOutgoingNodesToQueue(TransactionNode *transactionNode, int currentStackDepth,
+							List **toBeVisitedNodes)
+{
+	ListCell *currentWaitForCell = NULL;
+
+	/* as we traverse outgoing edges, increment the depth */
+	currentStackDepth++;
+
+	/* prepend to the list to continue depth-first search */
+	foreach(currentWaitForCell, transactionNode->waitsFor)
+	{
+		TransactionNode *waitForTransaction =
+			(TransactionNode *) lfirst(currentWaitForCell);
+		QueuedTransactionNode *queuedNode = palloc0(sizeof(QueuedTransactionNode));
+
+		queuedNode->transactionNode = waitForTransaction;
+		queuedNode->currentStackDepth = currentStackDepth;
+
+		*toBeVisitedNodes = lappend(*toBeVisitedNodes, queuedNode);
```

Comment:
Why is that? `D` won't be in the `deadlockPath` that we form. `D` would be removed from the `transactionNodeStack` given that `B` and `D` would have the same `stackDepth` and we do the following: `transactionNodeStack[currentStackDepth] = currentTransactionNode;`

If `D` isn't in the `deadlockPath` it is not going to be killed at all.

Also, the deadlock path would be `A->B>C->A`

Reply:
Oh, wait. I see now what you mean. That's unfortunate that we use `lappend` instead of `lcons` in `PrependOutgoingNodesToQueue()`.

(Well, I'm pretty sure that I've used `lcons` at some point, but, I cannot remember why it's `lappend` now. Let me think some more before opening the PR for the fix)

Reply:
FWIW this was reported as a new issue in #1960 and fixed in #1961.